### PR TITLE
serial: rework serial user interfaces

### PR DESCRIPTION
**Pull request by hereischen**
Signed-off-by: Haotong Chen <hachen@redhat.com>
id: 1482817

### CODE REVIEW

File: virttest/session.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            _func = getattr(self._console, func)
+            if not _func:
+                raise AttributeError("%s has no attribute %s." %
+                                     (self._console, func))
+            return _func(*args, **kwargs)
+        finally:
+            self._lock.release()
+
+
+class Session(object):
```
Comment (luckyh) [inline virttest/session.py]: How about `ConsoleSession`?

File: virttest/session.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    """
+    A wrapper for cummunicating with console.
+
+    For more detailed function call information,
+    you may need to refer to aexpect.clinet.ShellSession().
```
Comment (luckyh) [inline virttest/session.py]: Typo.

File: virttest/session.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    """
+    def __init__(self, manager):
+        self.manager = manager
```
Comment (luckyh) [inline virttest/session.py]: Should be a private member.

File: virttest/session.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        self._is_closed = False
```
Comment (luckyh) [inline virttest/session.py]: Suggest `__closed`.

File: virttest/session.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    def cmd_output(self, *args, **kwargs):
+        if self._is_closed:
+            raise RuntimeError("%s is closed." % self)
```
Comment (luckyh) [inline virttest/session.py]: Suggest:
```python
def __verify_not_closed(self):
    if self.__closed:
        raise RuntimeError("%s is closed." % self)

def cmd_output(self, *args, **kwargs):
    self.__verify_not_closed()
    ...
```

Comment (luckyh) [inline virttest/session.py]: Better to implement `__repr__` if needs to reference `self` here.

File: virttest/session.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    def cmd(self, *args, **kwargs):
+        if self._is_closed:
+            raise RuntimeError("%s is closed." % self)
+        return self.manager.execute(self.cmd.__name__,
+                                    *args, **kwargs)
+
+    def close(self):
+        if self._is_closed:
+            raise RuntimeError("%s is closed." % self)
+        self._is_closed = True
```
Comment (luckyh) [inline virttest/session.py]: Missing interface `is_responsive`.

File: virttest/virt_vm.py [hunk 7]
```diff
@@ -1242,7 +1245,10 @@ def serial_login(self, timeout=LOGIN_TIMEOUT,
         console.set_status_test_command(status_test_command)
 
         remote.handle_prompts(console, username, password, prompt, timeout)
-        return console
+
+        self.console_manager.set_console(console)
```
Comment (luckyh) [inline virttest/virt_vm.py]: This should be called in `create_*_console`.

File: virttest/virt_vm.py [hunk 8]
```diff
@@ -1242,7 +1245,10 @@ def serial_login(self, timeout=LOGIN_TIMEOUT,
...
+        session = self.console_manager.create_session()
+        return session
```
Comment (luckyh) [inline virttest/virt_vm.py]: In short `return self.console_manager.create_session()`.

File: virttest/virt_vm.py [hunk 9]
```diff
@@ -16,6 +16,7 @@
 from . import ppm_utils
 from . import data_dir
 from . import error_context
+from . import session
```
Comment (luckyh) [inline virttest/virt_vm.py]: Suggest the name `vm_console`.

File: virttest/virt_vm.py [hunk 10]
```diff
@@ -1242,7 +1245,10 @@ def serial_login(self, timeout=LOGIN_TIMEOUT,
         console.set_status_test_command(status_test_command)
 
         remote.handle_prompts(console, username, password, prompt, timeout)
```
Comment (luckyh) [inline virttest/virt_vm.py]: IMHO, the `create_console()` method should not be called any more in this function, and all the setup steps (including `console.set_linesep`, `console.set_status_test_command`, `remote.handle_prompts` ...) should be wrapped in `ConsoleManager`.

File: virttest/session.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+class NoConsoleError(Exception):
+
+    def __init__(self, *args):
+        Exception.__init__(self, *args)
+
+
+class ConsoleBusyError(Exception):
+
+    def __init__(self, *args):
+        Exception.__init__(self, *args)
```
Comment (luckyh) [inline virttest/session.py]: Suggest the following structure:
```python
class ConsoleError(Exception):
    ...

class NoConsoleError(ConsoleError):
    pass

class ConsoleBusyError(ConsoleError):
    pass

...
```

File: virttest/session.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        del state["_lock"]
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self._lock = threading.Lock()
+
+    def create_session(self):
```
Comment (luckyh) [inline virttest/session.py]: Docstring.

File: virttest/session.py (hunk 11 above)
Comment (hereischen) [reply virttest/session.py]: This commit is updated, thanks

File: virttest/vt_console.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+        self._console.set_status_test_command(self._status_test_command)
+        remote.handle_prompts(self._console, username, password,
+                              self._prompt, timeout)
+
+    def create_session(self, username, password, timeout):
+        """
+        Return a console session with itself as the manager.
+
+        :param timeout: Time (seconds) before giving up logging into the guest.
+        :param virtio: is a console virtio console.
```
Comment (YanhuiMa) [inline virttest/vt_console.py]: Pls update doc string

File: virttest/vt_console.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+
+from . import remote
+
+
+class ConsoleError(Exception):
+    """A base class for console errors."""
+
+    def __init__(self, *args):
+        Exception.__init__(self, *args)
```
Comment (YanhuiMa) [inline virttest/vt_console.py]: Can Line10 and Line11 be removed like class ConsoleBusyError? Is there special purpose?

File: virttest/vt_console.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+
+
+class NoConsoleError(ConsoleError):
+    """Raise when there is no console available for a session."""
+
+    pass
+
+
+class ConsoleBusyError(Exception):
```
Comment (luckyh) [inline virttest/vt_console.py]: Should inherit the base class.

File: virttest/vt_console.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        del state["_lock"]
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self._lock = threading.Lock()
+
+    def __login(self, username, password, timeout):
```
Comment (luckyh) [inline virttest/vt_console.py]: Should acquire the `_lock` before using console.

File: virttest/vt_console.py (hunk 13 above)
Comment (hereischen) [reply virttest/vt_console.py]: updated, thanks.

File: virttest/vt_console.py (hunk 15 above)
Comment (hereischen) [reply virttest/vt_console.py]: updated thanks

File: virttest/vt_console.py [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+    """Raise when session tries to communicate with a console in use."""
+
+    pass
+
+
+def lock(function):
+    """
+    Get the ConsoleManager lock, run the function, then release the lock.
+
+    :param function: Function to wrap.
```
Comment (CongLi) [inline virttest/vt_console.py]: doc string for 'return'

File: virttest/vt_console.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+                       prompt, username, password, timeout):
+        """
+        Return a console session with itself as the manager.
+
+        :param linesep: line separator for console
+        :param status_test_command: status test command for console
+        :param prompt: console prompt pattern
+        :param username: login username
+        :param password: login password
+        :param timeout: time (seconds) before giving up logging into the guest
```
Comment (CongLi) [inline virttest/vt_console.py]: doc string for return

File: virttest/vt_console.py (hunk 17 above)
Comment (hereischen) [reply virttest/vt_console.py]: It returns what `function` returns. I think we don't need that in doc string.

**Comment by CongLi** [comment]
do you have test results for this PR?

**Comment by CongLi** [comment]
Acked-by: Cong Li <coli@redhat.com>

**Comment by luckyh** [comment]
@ldoktor @balamuruhans call for your review, thanks!

**Comment by hereischen** [comment]
run with an acceptance test run
RESULTS    : PASS 123 | ERROR 0 | FAIL 2 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
None of these 2 failed case relate to this patch

File: virttest/vt_console.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        self.__login(linesep, status_test_command, prompt,
+                     username, password, timeout)
+        return ConsoleSession(self)
+
+    def set_console(self, console):
+        self._console = console
+
+    @lock
+    def proxy_call(self, func, *args, **kwargs):
+        """Proxy funcation call to call fucntions provided by a Console."""
```
[the remark this answers is not in the document]
Comment (hereischen) [reply virttest/vt_console.py]: @balamuruhans updated, thanks

File: virttest/vt_console.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+        self.__login(linesep, status_test_command, prompt,
+                     username, password, timeout)
+        return ConsoleSession(self)
+
+    def set_console(self, console):
+        self._console = console
+
+    @lock
+    def proxy_call(self, func, *args, **kwargs):
+        """Proxy function call to call fucntions provided by a Console."""
```
Comment (balamuruhans) [inline virttest/vt_console.py]: there is another %s/fucntions/functions

**Comment by luckyh** [comment]
Thanks all.

Let's merge this.

**Pull request merged**